controllers/etcd_controller.go

```diff
@@ -468,228 +463,6 @@ func (r *EtcdReconciler) getPodDisruptionBudgetFromEtcd(etcd *druidv1alpha1.Etcd
-				continue
-			}
-		}
-
-		// Fetch the updated statefulset
-		// TODO: (timuthy) Check if this is really needed.
-		sts := &appsv1.StatefulSet{}
-		if err := r.Get(ctx, types.NamespacedName{Name: claimedStatefulSets[0].Name, Namespace: claimedStatefulSets[0].Namespace}, sts); err != nil {
-			return nil, err
-		}
-
-		// Statefulset is claimed by for this etcd. Just sync the specs
-		if sts, err = r.syncStatefulSetSpec(ctx, logger, sts, etcd, values); err != nil {
-			return nil, err
-		}
-
-		// restart etcd pods in crashloop backoff
-		selector, err := metav1.LabelSelectorAsSelector(sts.Spec.Selector)
-		if err != nil {
-			logger.Error(err, "error converting StatefulSet selector to selector")
-			return nil, err
-		}
-		podList := &corev1.PodList{}
-		if err := r.List(ctx, podList, client.InNamespace(etcd.Namespace), client.MatchingLabelsSelector{Selector: selector}); err != nil {
-			return nil, err
-		}
-
-		for _, pod := range podList.Items {
-			if utils.IsPodInCrashloopBackoff(pod.Status) {
-				if err := r.Delete(ctx, &pod); err != nil {
-					logger.Error(err, fmt.Sprintf("error deleting etcd pod in crashloop: %s/%s", pod.Namespace, pod.Name))
-					return nil, err
-				}
-			}
-		}
-
-		sts, err = r.waitUntilStatefulSetReady(ctx, logger, etcd, sts)
-		return sts, err
-	}
-
-	// Required statefulset doesn't exist. Create new
-	sts, err := r.getStatefulSetFromEtcd(etcd, values)
-	if err != nil {
-		return nil, err
-	}
-
-	err = r.Create(ctx, sts)
-
-	// Ignore the precondition violated error, this machine is already updated
-	// with the desired label.
-	if err == errorsutil.ErrPreconditionViolated {
-		logger.Info("StatefulSet %s precondition doesn't hold, skip updating it.", "statefulset", kutil.Key(sts.Namespace, sts.Name).String())
-		err = nil
-	}
-	if err != nil {
-		return nil, err
-	}
-
-	sts, err = r.waitUntilStatefulSetReady(ctx, logger, etcd, sts)
-	return sts, err
-}
-
-func getContainerMapFromPodTemplateSpec(spec corev1.PodSpec) map[string]corev1.Container {
-	containers := map[string]corev1.Container{}
-	for _, c := range spec.Containers {
-		containers[c.Name] = c
-	}
-	return containers
-}
-
-func clusterScaledUpToMultiNode(etcd *druidv1alpha1.Etcd) bool {
-	if etcd == nil {
-		return false
-	}
-	return etcd.Spec.Replicas != 1 &&
-		// Also consider `0` here because this field was not maintained in earlier releases.
-		(etcd.Status.Replicas == 0 ||
-			etcd.Status.Replicas == 1)
-}
-
-func (r *EtcdReconciler) syncStatefulSetSpec(ctx context.Context, logger logr.Logger, ss *appsv1.StatefulSet, etcd *druidv1alpha1.Etcd, values map[string]interface{}) (*appsv1.StatefulSet, error) {
-	decoded, err := r.getStatefulSetFromEtcd(etcd, values)
-	if err != nil {
-		return nil, err
-	}
-
-	if reflect.DeepEqual(ss.Spec, decoded.Spec) {
-		return ss, nil
-	}
-
-	ssCopy := ss.DeepCopy()
-	ssCopy.Spec.Replicas = decoded.Spec.Replicas
-	ssCopy.Spec.UpdateStrategy = decoded.Spec.UpdateStrategy
-
-	recreateSTS := false
```

Comment:
Earlier we supported the recreation of the StatefulSet which was for instance required if the cluster was scaled up and the `spec.ServiceName` was changed in that scope (see following lines). I see that this logic has been removed but not added to the component. Am I missing something?

Reply:
Some fields are immutable, so you need to re-create the entire `StatefulSet` if the fields are changed for the supported use-cases.

Reply:
What I see that the recreation was called if a claimed statefulset was needed to be updated with immutable fields. But we are not claiming statefulset anymore. So, why do we need to update immutable fields of statefulset? Is there any need of claiming an existing statefulset? I may be missing something here

Reply:
Okay. I think, I got it. Will address this in next update of the PR shortly. Thanks.

Reply:
I don't see a pure relation to the claim logic. It is required to support:
- Changing `statefulset.spec.selector` (see [here](https://github.com/gardener/etcd-druid/blob/009a00f84103d61214e07d37db6a7030be15047e/controllers/etcd_controller.go#L613-L615)). Maybe this case is not very important anymore, esp. when we drop the claim logic.
- Changing `statefulset.spec.serviceName` (see [here](https://github.com/gardener/etcd-druid/blob/009a00f84103d61214e07d37db6a7030be15047e/controllers/etcd_controller.go#L617-L623)). Please see the comment in the code why this is required and important to support.

Reply:
Addressed